constraints/src/main/java/gov/nasa/jpl/aerie/constraints/time/IntervalMap.java

```diff
@@ -0,0 +1,491 @@
+        interval(bounds.end, bounds.endInclusivity.opposite(), Duration.MAX_VALUE, Inclusive)
+    ));
+  }
+
+  public void clear() {
+    this.segments.clear();
+  }
+
+  /**
+   * Authored by Jonathan
+   * Maps intervals and the gaps between them in IntervalMap intervals to new values following some function transform
+   *  which converts the old values (or nulls, in the case of gaps) to new values.
+   *
+   * @param transform the function to apply to each value of intervals (or the gaps between!), mapping its original value type V to another value type R
+   * @return a new IntervalMap with newly mapped values
+   * @param <R> The new value type that the returned IntervalMap's intervals should correspond to
+   */
+  public <R> IntervalMap<R> map(final Function<Optional<V>, Optional<R>> transform) {
+    return this.contextMap((v, i) -> transform.apply(v));
+  }
+
+  public <R> IntervalMap<R> contextMap(final BiFunction<Optional<V>, Interval, Optional<R>> transform) {
+    final var segments = new ArrayList<Pair<Interval, R>>();
+
+    var previous = alg.bottom(); //in the context of Windows, a interval at Duration.MIN; a minimum value when computing gaps at the next step
+    for (final var segment : this.segments) {
+      //previous might be ----TT---
+      //segment might be  -------F-
+      //gap is then       ------+--
+      final var gap = alg.intersect(alg.upperBoundsOf(previous), alg.lowerBoundsOf(segment.getKey()));
+
+      //we apply the transform to the gap (if it has contents)
+      //currently we pass in an Optional.Empty to the function so it can handle a gap that isn't in our intervals, in case
+      //  that should actually be mapped to a value (i.e. turn null into false for whatever reason). Then the return value
+      //  of that, which is now Optional<R>, is checked for value, if it has any, add that and the new R value
+      if (!alg.isEmpty(gap)) {
+        transform.apply(Optional.empty(), gap).ifPresent($ -> segments.add(Pair.of(gap, $)));
+      }
+
+      //apply the transform to the actual segment
+      //returns an Optional<R>, check if it has value (it might not, if for example the transform maps Optional<V> where
+      //  the value in that optional is true to null/Optional.empty(), in which case it won't have value and we don't
+      //  wish to add to the map), and then if so add to the map
+      transform.apply(Optional.of(segment.getValue()), segment.getKey()).ifPresent($ -> segments.add(Pair.of(segment.getKey(), $)));
+
+      previous = segment.getKey();
+    }
+
+    {
+      //check the final gap
+      final var gap = alg.upperBoundsOf(previous);
+      if (!alg.isEmpty(gap)) {
+        transform.apply(Optional.empty(), gap).ifPresent($ -> segments.add(Pair.of(gap, $)));
+      }
+    }
+
+    return new IntervalMap<>(this.alg, segments);
+  }
+
+  //Jonathan's implementation
+  public static <V1, V2, R>
+  IntervalMap<R> map2(
```

Comment:
That was exclusively @pranav-super 's work, I didn't change it from what I received from him. I'll dig through and see what the changes do.

Reply:
I'm not sure anymore. The old implementation broke in a way that I couldn't figure out when I removed `IntervalAlgebra.horizon` and `alg.bottom()`, so I rewrote it. Its a little uglier, but the general algorithm is simpler: one iteration of the loop per call to `transform.apply`, in chronological order.

I tried taking your old prototype and using it instead, with modifications to remove `alg.bottom()`. It hit an infinite loop on one of the tests because it stops advancing the intervals, which is a problem that I encountered too when I rewrote it, but I don't remember the exact edge case.

Reply:
I'm not sure it matters, but can you remind me why `alg.bottom()` needed to be removed? Isn't it just the empty interval? If that makes the implementation easier, it seems worth having.

Which test(s) broke? I can try to take a look too.

Reply:
`alg.bottom()` was `Interval.at(horizon.start)`.

I'm not concerned about the different possible implementations of map2. They all have the same space and time complexities, and Pranav wrote many, many tests for it. For map2 at least, as long as the tests all pass I'm satisfied.

map2emptyAll gets into a loop.

Reply:
> `alg.bottom()` was `Interval.at(horizon.start)`.

Ah. Well, that's not the "bottom" element of intervals. That element should be the identity for `unify`, which would be the empty interval.

> I'm not concerned about the different possible implementations of map2. They all have the same space and time complexities, and Pranav wrote many, many tests for it. For map2 at least, as long as the tests all pass I'm satisfied.

I agree... it's just a very, very complex block of code. I worry about maintainability here, not efficiency.

> map2emptyAll gets into a loop. So does unsetAllSparseInMiddle and probably some others.

Ok, thanks.

Reply:
Oh my bad, `bottom` was `Interval.between(start, Exclusive, start, Exclusive)`

Reply:
🤔 That checks out, then :+1:

Reply:
`Interval.between(0, -1)` is also an empty interval (and already exists via `Interval.EMPTY`), so I'm still not sure this explains why `alg.bottom()` needed to be avoided. 🤔

Reply:
it didn't need to be removed, but now that its gone I don't see a reason to put it back. If its location at the horizon start mattered, it is not longer valid because `horizon` is gone. If its location didn't matter, you can just use `Interval.EMPTY`.

Reply:
Sorry, it sounded like its removal was a big reason you had to reimplement the method. If that's not the case, it doesn't matter.

Reply:
Ohhh no it was something about the truncating behavior at the bounds that I couldn't figure out. I wanted something easier to reason about.

Reply:
Ok! 👍 I'll sit on my misgivings until we switch to a transition-based representation, which should help this anyway.

Reply:
Also I wasn't aware that empty intervals should be identity in unify. Looking at `Interval.unify` it doesn't look to me like it actually behaves that way. So when I tried using your prototype, I used `Optional<Interval> accInterval` instead of making the default `accInterval = Interval.EMPTY` or something similar. And it still hit the infinite loop.

Reply:
> Looking at `Interval.unify` it doesn't look to me like it actually behaves that way.

Well, that... uh... that's pretty clearly a bug, honestly. 👀 It looks like `unify` was implemented symmetrically with `intersect`, when the extremal intervals actually make for very different edge-case handling.

Reply:
It just needs two guards at the top to be fixed. That's definitely not the intended behavior.

```java
    if (x.isEmpty()) return y;
    if (y.isEmpty()) return x;
```

Reply:
There's also no check that the intervals overlap or meet. Without that this seems more like a hull operation. Was that the intended behavior?

Reply:
Yes, it has to be a hull operator -- the point-wise union of two intervals is not, in general, a union. The "least upper bound" of two intervals must give the *smallest* interval *containing* both given, which here is the hull.

Reply:
This is all stuff I should have written down in javadocs when I first implemented it. Mea culpa, I came back for a reason.

Reply:
I've changed the name to `hull` and added the checks for empty intervals 👍

Reply:
Well, a `hull` operator isn't usually binary, so I think that name is confusing in a different way. It also loses the intentional symmetry with `intersect`; these two operators plus top/bottom define a bounded boolean lattice, which just means that they behave like AND and OR in logical settings.

If you don't like `unify`/`intersect`, you could go with `leastUpperBound` and `greatestLowerBound`, or `lub`/`glb`, or `and`/`or`, or `join`/`meet` (but even mathematicians get those two mixed up). But `unify`/`intersect` seemed like the most natural choice at the time.

Reply:
(This correspondence is *why* the class was called `IntervalAlgebra`, by the way.)

Reply:
I think its misleading that `unify` doesn't perform the `union`. But I don't care that much, I'll change it back.

Reply:
Yeah. The "right" names are "join" and "meet" (see [here](https://en.wikipedia.org/wiki/Boolean_algebra_(structure))), but I figured those names would be a step too far for most readers.

Sorry, and thanks. 👍